### ISSUE TITLE
Non Qwerty\En US Input

### ISSUE THREAD

**Issue by Kethku** [neovide/neovide#445]
Right now keyboard layouts are handled by manually listing out bindings given key presses and modifiers. This is very broken.

Luckily there is movement here which should fix this issue. https://github.com/rust-windowing/winit/pull/1788

Once this lands and looks stable enough. I'm going to pull in some version of winit which uses it, and rewrite our keyboard input handling to use that. Most of the hard work was already done thanks to @j4qfrost, so now its a waiting game. Looks like its moving very quickly though so I'm hopeful we will have motion on this soon.

For now I'm going to comb through our issues looking for any that have to do with keyboard input handling and close them as this should fix them or at least make a fix possible.

**Comment by pickfire** [comment]
Note that IME support is different from keyboard layout, IME uses an external program unlike using a different layout/variant.

**Comment by Kethku** [comment]
Agreed. However IME support in sdl2 is not good either. I think at this point supporting multiple windowing systems is not the way to go, so I'm going to be moving to switch to winit only once the upstream changes land. At that point, we can work with them to get IME support working for real on the relevant platforms.

**Comment by Kethku** [comment]
https://github.com/Kethku/neovide/pull/506 a PR has been contributed targetting the new winit api. It currently only works on windows, but once upstream support for linux and mac lands, we should be in very good shape.

**Comment by Kethku** [comment]
https://github.com/rust-windowing/winit/pull/1890 looks like an upstream macos PR has been created. Very excited

**Comment by p00f** [comment]
I've set my window manager to swap escape and caps lock but it doesn't work in neovide (they are not swapped), will this fix my issue too?

**Comment by pickfire** [comment]
> I've set my window manager to swap escape and caps lock but it doesn't work in neovide (they are not swapped), will this fix my issue too point_rightpoint_left ?

I did the same but IIRC I don't have this issue, I only have an issue with dvorak keys when used with other combination keys, I even swap my left ctrl and left alt but I think I don't have an issue with that either. Maybe I did but I don't recall.

**Comment by Kethku** [comment]
I believe this would fix both of your problems @p00f @pickfire. At the very least it would unblock addressing those issues.

**Comment by Kethku** [comment]
https://github.com/rust-windowing/winit/pull/1788 PR one of 3 has merged! Shouldn't be long now!!!

**Comment by Pebor** [comment]
12 days ago 
![image](https://user-images.githubusercontent.com/32438954/117420096-89752400-af1d-11eb-9847-54a8bf4f6509.png)
But hype!

**Comment by Kethku** [comment]
Note, we need mac and linux support to merge first

**Comment by Kethku** [comment]
@PyGamer0 seems possible.

**Comment by Kethku** [comment]
Linux: https://github.com/rust-windowing/winit/pull/1932
Mac: https://github.com/rust-windowing/winit/pull/1890

There are now PRs for both mac and linux implementations. Making good progress.

**Comment by ewen-lbh** [comment]
For those that need a workaround while this is getting fixed, you can input any character into the current text buffer by typing the following (from Normal mode):

```
:r!echo 'YOUR CHAR(S)'
```

Aliasing echo to sth. like `e` might make the process slightly less annoying

**Comment by p00f** [comment]
> I've set my window manager to swap escape and caps lock but it doesn't work in neovide (they are not swapped), will this fix my issue too point_rightpoint_left ?

this seems to have been fixed now, nice

**Comment by PyGamer0** [comment]
> I am using uncap, but it doesn't work.. Is this related to this issue also?

this is fixed on opengl branch

**Comment by TobiasDev** [comment]
Hello! 
Will this merge (once a new release comes), also solve the issue with having a Swedish keyboard? Currently when I try to do several different keys it gives me a different character than I expect. E.g. [ turns into something different than it should.

**Comment by Kethku** [comment]
It should

**Comment by fredizzimo** [comment]
I saw that rust-windowing/winit#1788 is now merged, so I compiled the latest from source on Windows. However, the input still appears to be broken at least with the Swedish layout.

Regular characters like `åäö` works fine, and so do most special characters. I did not find any broken ones in combination with shift, but alt-gr combinations have a lot of problems.

For example `alt-gr+2` should produce `@`, but it just acts as `ESC`.  On the other hand `alt-gr+3` correctly produces `£` for example, I don't really see any pattern, just that it's consistent.

**Comment by fredizzimo** [comment]
Ah, sorry, I did not read everything properly, it needs to be integrated into neovide too

**Comment by Kethku** [comment]
Yeah we need to do integration work on our side first. Theres an open PR which attempts this, but I think there will be some more fiddling on top of that.

**Comment by Kethku** [comment]
https://github.com/Kethku/neovide/issues/691 tagging here. I think it will be necessary to switch to scancodes for linux machines while we wait for the upstream PRs to merge.

Alternatively I may try to make a fork which pulls both the linux and mac versions of the keyboard changes

**Comment by Kethku** [comment]
https://github.com/Kethku/neovide/pull/736

This integrates the new keyboard api by depending on my merged forks of the upstream keyboard pr branches.

This brings us up to the current level of capability. I would appreciate if folks here could try this PR and let me know:

1. If keyboard handling is different than you would expect
2. What OS you are using, and what keyboard format do you have set
3. What keys exactly did you press
4. What happened in the editor
5. What did you expect to happen

This way I can iterate on the PR a bit. I think the new PRs will let us do something that works how people expect, but it will probably need some fiddling

**Comment by TobiasDev** [comment]
I can't wait for a new build with this! I love Neovide, but since my default keyboard is a Swedish one I can't use it for everything yet. :D I still do when I code though since I use the US layout for that. But once this is integrated I will be able to use it for all! :)

Thanks for all the good work!

**Comment by last-partizan** [comment]
I tried that branch, and I found no issues (Archlinux Gnome/Wayland, standard 105 keys keyboard layout).

(I still cannot use it daily becouse of rendering issues in #723, but at first glance all keys/combinations working as expected)

**Comment by cljoly** [comment]
I’ve tried `keyboard-rework` and it seems to work.
1. The keyboard handling is as I would expect
2. I’m using Arch Linux with wayland ([sway](https://swaywm.org/)). It is what sway calls [`French (BEPO)`](https://bepo.fr/wiki/Accueil), which is dvorak-like keyboard for French.
3. As you can imagine given the keyboard layout, I have a lot of things remapped. All the below worked flawlessly:
   - directional keys: `h`, `k`… are remapped to `r`, `s`…
   - `w` is remapped to `é`
   - some custom shortcut (in normal mode) like `à`, `ê` or `<space>à`
   - typing letters with diacritics like `é`, `ç`… in insert mode
4. & 5. I expected the editor to type the letter I would get in my terminal or web browser and it is what happened!

This is very promising, thanks a lot for your work on this branch!

**Comment by danielrparks** [comment]
1. If I press a key with the shift key held, and then release the shift key, neovide doesn't register the shift key release, and every character that I type afterwards will be interpreted as capitalized.
2. Arch Linux with Sway (Wayland). Plain `dvorak`.
3. `ihello world<S-1> the special k... wait why is everything stuck in caps<ESC>ohow do I exit<ESC><S-;>q<S-1><Return>`
4. Actual:  
![Screenshot from 2021-06-20 08-19-32](https://user-images.githubusercontent.com/42898924/122675709-4b1b8780-d1a0-11eb-83f1-b7d8ca8cc1c1.png)
5. Expected:  
![Screenshot from 2021-06-20 08-21-21](https://user-images.githubusercontent.com/42898924/122675763-8ae26f00-d1a0-11eb-98ed-a8ecfc9b3f80.png) (exits)

**Comment by cljoly** [comment]
@danielrparks I have the same bug as you, except that it happens for all modifier and that I can’t reproduce it as consistently (pressing a modifier rarely triggers it).

**Comment by Kethku** [comment]
Ok that's a valuable hint. I have some ideas

**Comment by Kethku** [comment]
@cljoly, @danielrparks I pushed a small change that I think should address the modifiers issue. Can you try again?

**Comment by danielrparks** [comment]
That's fixed the shift-stuck bug, but now I found a new one.

I use [auto-pairs](https://github.com/jiangmiao/auto-pairs) to automatically insert closing symbols. In this build, auto-pairs only works on symbols that are not shifted, like ``, [], ''. It does not automatically complete pairs for (), {}, "", <>.

**Comment by ccakes** [comment]
> I’ve tried the branch `keyboard-rework` and it seems to work.

:+1: Extra data point, works here too. Arch, sway, US layout (altgr-intl variant). Thanks for the awesome work!

**Comment by ouuan** [comment]
> Agreed. However IME support in sdl2 is not good either. I think at this point supporting multiple windowing systems is not the way to go, so I'm going to be moving to switch to winit only once the upstream changes land. At that point, we can work with them to get IME support working for real on the relevant platforms.

Just want to make sure, is the progress of IME support also being tracked here? I think they are two different issues. The keyboard layout problem seems to be nearly fixed, but I don't know the status of IME support. Perhaps it will be better if we can track the progress of IME support in a different issue.

**Comment by last-partizan** [comment]
After #723 is fixed, i'm tried to use this branch again, and found another bug:

When i press `%` it moves between matching braces, this is works.

But, in raw neovim it also moves between opening/closing html tags, this part is now broken.

**Comment by ITHackerstein** [comment]
Hi, I tried out the `keyboard-rework` branch, and I've had some interesting results trying out a bunch of key combinations.
I noticed is that when I maximize the Neovide window using `Win`+`UpArrow` I get some weird character sequences. Here are some examples for some key combinations:
* `AltGr` `Shift` `è` (should output `{`) --> `<S-D-{>`
* `AltGr` `Shift` `+` (should output `{`) --> `<S-D-}>`
* `Tab` (should output a tab character) --> `<D-Tab>`
* 'Alt' 'Tab' (should switch to another window) --> `<M-D-Tab>` and switches to normal mode.

If I instead don't maximize the window or I maximize it using the title bar button I get the correct keys.

OS: Ubuntu 20.04
Keyboard layout: [QWERTY Italian](https://it.wikipedia.org/wiki/File:Italian_Keyboard_layout.svg)

**Comment by danielrparks** [comment]
In the same vein as what @last-partizan noticed, using `%` to move between begin/end environment statements in vimtex is also not working. Instead, it tries to move between the curly braces.

For example, instead of moving from `|\begin{enumerate}` (pipe represents the cursor) to `\end{enumerate|}`, it moves from `|\begin{enumerate}` to `\begin{enumerate|}` (it moves to the curly brace on the same line).

**Comment by fredizzimo** [comment]
I tested it on Windows with the Swedish keyboard layout, and the dead keys don't seem to work correctly. But it's a big improvement from the previous version, since everything else seems to work.

More specifically dead keys followed by a space or an unsupported character don't work. For example

`¨` followed by `e` correctly outputs `ë`
`¨` followed by `<space>` incorrectly outputs `<space>`, but it should output `¨` (notice that the space should not be output in this case)
`¨` followed by `d` incorrectly outputs just `d`, it should be `¨d`
`¨` followed by another `¨` incorrectly outputs nothing, it should be `¨¨`

All the dead keys `` ¨^~´` `` behave the same way

Especially the second case with space is very important since that's the only way to ouptut common symbols like `` `~^ ``, althought I suspect that some people who don't know better might rely on the last case and press the dead key twice and then delete the extra character.

**Comment by cljoly** [comment]
> That's fixed the shift-stuck bug, but now I found a new one.
> 
> I use [auto-pairs](https://github.com/jiangmiao/auto-pairs) to automatically insert closing symbols. In this build, auto-pairs only works on symbols that are not shifted, like ``, [], ''. It does not automatically complete pairs for (), {}, "", <>. On standard terminal neovim, it works as expected.

Same for me on those two points.

**Comment by cljoly** [comment]
> I tested it on Windows with the Swedish keyboard layout, and the dead keys don't seem to work correctly.
@fredizzimo I have the same behavior on Sway (wayland) with Arch Linux.

**Comment by kilpkonn** [comment]
For me `Ctrl+Shift+V` paste seems to be not working in neovide
Also noticed that when switching between panes in xmonad with `Super+nr` neovide detects it as input while nevide in Alacritty terminal does not.
OS: Arch with xmoad

**Comment by Kethku** [comment]
@fredizzimo @cljoly I suspected this would be an issue with dead keys. The problem is that I don't know what the correct behavior should be for handling both bindings in normal mode and text input in input mode. I think the PRs upstream give us tools which should make the problem tractable, but I don't know enough about common keyboard patterns with non qwerty keyboards to be confident in what changes are necessary.

A couple questions (I'll have more):
1. What should happen you configure a keybinding to a deadkey? Is that something anybody does? Should it be a no-op?
2. Should normal mode keyboard handling act differently from insert mode? We could pipe the current mode into the window system, but I would need to know what people would expect the difference to be.
3. Its not clear to me how modifiers should get sent to neovim. My understanding of keyboard layouts is that sometimes a modifier is "consumed" in order to modify the inserted text for a given key press. I don't know if I have the ability to tell whether a particular modifier has been used up in this way given the api I have from winit. Is it ok to still pass the modifier even if it has modified the associated key? Maybe I'm completely confused haha

Answers to these questions would go a long way toward helping me know what the correct implementation of this stuff should be. I'm very much not an expert in this area, so fixing this would require some time and commitment to help me understand. Alternatively I would be happy to assist somebody if they want to dive in. Either way works for me

**Comment by fredizzimo** [comment]
@Kethku,

Yes, dead keys can be a bit hard to understand until you have actually used them. But the behaviour we are looking for here is quite simple. Normally the terminal handles the dead keys, and Vim will only see the end result, so that's what should happen here too.

So to answer your questions,
1. Keybindings to deadkeys are perfectly normal, the default binds `^` (go to first non blank on the line), and the backtick `` ` `` for marks. To use them you would have to press the the dead key followed by space. So to go to the start of the line I would press `shift+^` and then `<space>`. I don't know how common it is, but it's also possible to map things like `nmap ë ihello`, which would enter insert mode and type hello if I press `¨` followed by `e`. 

   It's also possible to map the other way around, and I guess many people do that for example `map ö ^`, so that they can press the dedicated `ö` key instead.

   Personally I have a programmable QMK keyboard, so I work around the problem by having a keymap that sends the space combination by default, with the actual dead keys are on another layer. 
 2. All VIM modes should behave exactly the same way, since the dead key processing is the first thing that should be done.
 3. For dead keys, the modifier should be handled separately for the actual dead key and the following letter. So I can for example press `altgr+~` and then `shift+a` for `Ã`. I'm also fairly sure that you should pass the modifiers that were held down at the moment the second key was pressed to Vim. 

A final note. Most keyboard layouts have simple dead keys with just two key presses, but some layouts have chainable dead keys, where you can press two or more dead keys in any order followed by the key to modify. I don't know how popular those are, and if they need to be supported, but the principle for supporting them is the same, wait for the final modified character and pass that to Vim along with the modifiers.  But it can go even beyond that, for example, I know that the Vietnamese keyboard layout has a slightly more complex variant of the dead keys, and that the Japanese and Chinese input methods are completely different.

**Comment by cljoly** [comment]
> I don't know how common it is, but it's also possible to map things like nmap ë ihello, which would enter insert mode and type hello if I press ¨ followed by e.

I have a few mappings on things with dead-keys, it works quite well in the terminal. I’m not saying it is hugely popular though.

> I'm also fairly sure that you should pass the modifiers that were held down at the moment the second key was pressed to Vim.

That’s the behavior I have with Alacritty.

**Comment by Kethku** [comment]
Ok I've made another stab at better dead key handling: https://github.com/Kethku/neovide/tree/keyboard-deadkey-handling

Can you try this and let me know if it works how you would expect? I don't have time now, but I will explain the logic later if that would help. The important logic is here: https://github.com/Kethku/neovide/blob/keyboard-deadkey-handling/src/window/window_wrapper/keyboard_manager.rs#L155-L170

**Comment by Kethku** [comment]
@danielrparks I think the keyboard-deadkey-handling branch may address your shift key-auto pair problem. Could you try again?

**Comment by danielrparks** [comment]
Everything seems to be working properly for me now, including auto-pairs and using % to move between html tags / latex environment begin & end. Note that I don't use dead keys, but I do use the compose key. Input using the compose key worked as expected, including custom mappings from `~/.XCompose`, and I was even able to map compose key combinations to neovim commands.
Thanks for putting so much effort into making sure neovide works for everyone!

**Comment by fredizzimo** [comment]
Thank you @Kethku,

The dead keys seem to work perfectly now, at least for the cases I can think of.

**Comment by clason** [comment]
Here's my (non-working) setup:

* macOS Big Sur (Intel)
* German keyboard layout

If I type `option-8` (which is `{` in my layout), I just get dropped out of insert mode. The relevant log events are
```
TRACE [neovide::channel_utils] ui_command Keyboard("<M-8>")
TRACE [neovide::redraw_scheduler] Next frame queued
TRACE [neovide::bridge::ui_commands] Keyboard Input Sent: <M-8>
```

**Comment by cljoly** [comment]
Thanks @Kethku, in every way I could think of, dead keys work for me as well, in particular the mappings on dead keys.

**Comment by Kethku** [comment]
@clason what happens in terminal vim for your use case. Alt in base vim escapes from insert mode. Its possible that your use case will need some setting for langmap :(

**Comment by clason** [comment]
Well, obviously in terminals it works as expected, or I would never have started using neovim :)

(To be honest, that's the first I heard that alt is not a modifier key but escapes insert mode on US keyboards...)

**Comment by danielrparks** [comment]
According to the wikipedia page linked by @clason, those additional characters are accessed with AltGr. I think the problem is that neovide is seeing it as plain Alt and not AltGr.

**Comment by clason** [comment]
Well, `AltGr` is the German name on the classical Windows keyboard. (I no longer have access to a Linux workstation, sorry.) 

On macOS it's just the `Option` key, either left or right (`M-` in the log above).

**Comment by danielrparks** [comment]
What I mean is that to work correctly, I think neovide needs to treat it like AltGr in that situation. I agree that this is probably a macOS-specific thing, as both Windows and Linux separate AltGr from Alt.

**Comment by clason** [comment]
Right, yes, that's pretty much what I meant, too ;)

**Comment by Kethku** [comment]
Ok so I think the thing I have to do now is get a version of the winit keyboard tester app compiled using the macos keyboard rework branch https://github.com/maroider/winit_keyboard_tester so that we can report the issue upstream

**Comment by Kethku** [comment]
Specifically we would report the issue here: https://github.com/rust-windowing/winit/pull/1890

**Comment by clason** [comment]
Specifically, you want to distinguish `Alt` and `Meta` on macOS.

**Comment by Kethku** [comment]
@clason I've commented on the relevant PR upstream. On windows holding the right alt and pressing 8 correctly reports { when using a german keyboard layout. So I think the mac version of the upstream keyboard api rework has a bug.

**Comment by Kethku** [comment]
https://github.com/rust-windowing/winit/pull/1890 to be specific, I commented here

**Comment by last-partizan** [comment]
I found another bug on Linux/Wayland:

In terminal neovim `Shift + Right Arrow` moves cursor to the end of the word, in neovide - it doesn't.

**Comment by danielrparks** [comment]
Interestingly, Ctrl + Arrow, which does the same, _does_ work.

**Comment by Kethku** [comment]
I've investigated and made a fix for the shift+arrow key issue. Also some awesome folks (@clason @smolck) have been helping me debug and fix macos specific issues.

Could folks here try again on main to see if any new issues have cropped up? I think we're getting close!

**Comment by clason** [comment]
At least on macOS, the latest commit behaves the same as in terminal neovim (`shift-right arrow` jumps to the _beginning_ of the next  word).

**Comment by danielrparks** [comment]
> Could folks here try again on main to see if any new issues have cropped up? I think we're getting close!

Everything is still working correctly for me, AFAICT.

**Comment by cljoly** [comment]
For me as well, everything still works correctly! 👏🏽

**Comment by last-partizan** [comment]
`Shift + Arrow` is fixed, i found no other issues today.

**Comment by Kethku** [comment]
OK i think this is fixed! Thanks for the help everyone!

**Comment by PyGamer0** [comment]
I think this can be unpinned now

**Comment by ewen-lbh** [comment]
Everything's working just fine, except that I can't use my compose key...

I set it with `setxkbmap -option compose:rwin`

**Comment by jsosulski** [comment]
Since  f0a46dbff9da43694ecff89a37bacd635788b653 I have some keybindings that use Shift in a sequence with <leader> key, which stopped working, e.g.:

```vim
map <Leader><tab> :bn<cr>
map <Leader><S-tab> :bp<cr>
```

The Binding for next buffer works, for previous buffer does not. However, for me this is not a high priority issue, as I tend to have few buffers open anyway and usually `:bn` cyclically.

**Comment by smolck** [comment]
@jsosulski what's your OS?

**Comment by tpict** [comment]
Apologies if this is the an issue with the upstream PR mentioned above–is it a known issue that Neovim doesn't receive the "meta" modifier at all after these changes? For example, `Alt + x` Bash shortcuts don't work inside a `term`. I'm on macOS and I'm honestly not sure what the correct behavior is. The built-in terminal app has a "use option as meta" preference for this reason.

**Comment by clason** [comment]
Yeah, I think that is the price to pay here -- you can't have `option` as both input and command modifier, and _not_ having the former makes the GUI unusable for people with certain keyboard layouts. 

Only thing I can think of is to turn this into an opt-out feature by adding a configuration option. Not ideal, but /shrug.

**Comment by smolck** [comment]
@tpict Right, as clason said, that's intended; I was under the impression that there wasn't really a way of doing alt or meta mappings normally in the TUI nvim, and so it wasn't an issue to break them to fix this issue, because they were never supposed to work (pressing `option+5` for example on US layout enters a character, ∞, and so can't really be used as a keybinding; or so I thought). 

If you don't use that macOS-specific functionality though, I could see how providing such a "option as meta" binding/setting could work/be useful. It would come with the caveat of not being able to type characters with option or option+shift as is usually the case on macOS, though.

**Comment by jsosulski** [comment]
@smolck I'm using KUbuntu 20.04 with Plasma/KWin 5.18 on Kernel 5.8.0 and nvim on nightly, i.e. 0.5.0

**Comment by tpict** [comment]
@clason @smolck Thanks both. I think the configuration option is the de facto solution considering it's what Apple did for Terminal.app. I have a bunch of Neovim mappings that use the `M-` modifier, so I'm definitely biased, but I don't think it's particularly uncommon to do so.

**Comment by smolck** [comment]
@tpict mind giving https://github.com/smolck/neovide/tree/alt-as-meta a test, with `cargo run --features alt-as-meta`? Should provide the functionality you're looking for I *think*.

**Comment by tpict** [comment]
@smolck Wow, thanks for the quick turnaround!

- `M-` Vim bindings work as expected
- `Alt + t` to swap words in a terminal works
- `Alt + e` for "edit current line in `$EDITOR`" in a terminal does _not_ work

I think `Alt + e` is behaving as a dead key?

**Comment by smolck** [comment]
@tpict Fixed now?

@jsosulski Mind giving https://github.com/smolck/neovide/tree/shift-tab-mappings a test? Should fix `<S-Tab>` mappings (seems to work for me from a quick test, at least).

**Comment by tpict** [comment]
@smolck almost–the shell receives the `Alt + e` event, but Neovim ignores the next keypress.

**Comment by smolck** [comment]
@tpict Hmm, so this is a more difficult issue . . . on US layout, pressing option+e results in the dead key `´`, which then waits for the next keypress to see if it should enter a character like `é`. I have no idea how to disable that.

Since the macOS terminal is written in Obj-C or Swift, it can probably do some OS tricks that disable the option functionality; regardless, I don't know how it achieves that functionality, so I can't try to port it to Rust.

Now, iTerm2 does have the same functionality and it's open-source, so I could look there, but I'm a little hesitant because it's licensed GPLv2 and I don't know if there are license issues with getting inspired by/using a derivative of the code it has to do that, so . . . kinda stuck right now.

An easy workaround is to just use a different key for the keybinding, but I know that's of course not ideal.

**Comment by smolck** [comment]
On second thought, actually I think this is an upstream issue/feature that winit needs to deal with: https://github.com/rust-windowing/winit/issues/768

**Comment by jsosulski** [comment]
@smolck Hm, the fix does not work for me. I tried both `<S-Tab>` and `<S-tab>` if that would make a difference. But at least in my shortcut with a leader sequence it does not work. My leader is space if that is relevant.

**Comment by Kethku** [comment]
Ok I was worried about this issue. I believe that most of the problems can be explained by an inconsistency between macos handling of keyboard inputs and windows/linux handling. With the amazing help of @smolck and @clason we iterated on the current approach and decided that meta on macos should be ignored for certain keypresses in favor of handling keyboard layouts better. I'm now convinced that the solution is just to fix the issue in the upstream PRs. I've created an issue which captures the disconnect https://github.com/rust-windowing/winit/pull/1890#issuecomment-869165792. However it appears that movement on that PR is understandably slow.

I don't have the cycles at the moment to dig in to the winit codebase right now and understand whats going on from their side. Especially since I don't have mac hardware to test with. So I think the solution for the moment is to wait to see if movement happens on their side.

@smolck @clason I'm adding you two as collaborators to the project (no responsibilities here, just trying to give flexibility). I'm going on a couple day trip so I won't be around for a bit to more directly help out here. When I get back though I'm going to take another stab at this.

**Comment by smolck** [comment]
> @smolck Hm, the fix does not work for me. I tried both `<S-Tab>` and `<S-tab>` if that would make a difference. But at least in my shortcut with a leader sequence it does not work. My leader is space if that is relevant.

@jsosulski Hmm, that's weird; `:map <Leader><S-Tab> <cmd>lua print('works')<cr>` works fine for me; my leader is different from yours  though, `;`, maybe that makes a difference? I don't see why it would though . . . could also be because I'm on macOS, but I *really* doubt that . . .

> @jsosulski your issue is a different problem though that should be addressable. Just need to find the time to figure it out. We may need to special case text characters that don't have a capital variant (such as tab) and send shift for those characters even when we don't for alphanumerics.

That's exactly how I'm fixing that issue with https://github.com/smolck/neovide/commit/4a9377565d1d6ccf97aa1b9ad8b71b6bf1489856, just have to see why it's only working locally if that's indeed the case. I can't think of any other keys like tab that aren't changed with shift but I could add those if there are any.

> @smolck @clason I'm adding you two as collaborators to the project (no responsibilities here, just trying to give flexibility). I'm going on a couple day trip so I won't be around for a bit to more directly help out here. When I get back though I'm going to take another stab at this.

Awesome, thank you!

Also, @Kethku maybe a new issue should be made for these remaining problems? Obviously it works/is easier to just keep things here, but the issue in the title, trying to input text with other keyboard layouts, *is* fixed; all that remains are weird edge cases with keybindings, it seems.

**Comment by jsosulski** [comment]
@smolck Hm for what it's worth, it seems the issue was with my vimrc. With a blank one containing only the one <S-tab> mapping this fix works. Thank you! On to debugging my vimrc

**Comment by last-partizan** [comment]
I found one more issue, when entering command, `Shift+Tab` doesn't go backwards in list.

![image](https://user-images.githubusercontent.com/301015/125090031-95c15d00-e0d7-11eb-9215-2f7dfff2e747.png)

**Comment by ewen-lbh** [comment]
My report got a bit buried under others, so I'm re-stating it in case it has gone unnoticed:

> Everything's working just fine, except that I can't use my compose key... (it seems like it just ignores the `rwin` press and inserts the compose sequence's characters verbatim)
> 
> I set it with `setxkbmap -option compose:rwin`

**Comment by gshpychka** [comment]
Please advise if it's related - my mouse scroll wheel is not working in neovide.

**Comment by last-partizan** [comment]
@gshpychka no, it's not related.

You should configure your mouse, like this `set mouse=nv` (also, see `:help mouse`)

**Comment by ewen-lbh** [comment]
> My report got a bit buried under others, so I'm re-stating it in case it has gone unnoticed:
> 
> > Everything's working just fine, except that I can't use my compose key... (it seems like it just ignores the `rwin` press and inserts the compose sequence's characters verbatim)
> > I set it with `setxkbmap -option compose:rwin`
> 
> Appart from this, I've been using neovide with the fix applied for almost a week and everything else works perfectly.

Seems like the bug is gone, I can use compose keys just fine now.

**Comment by lijaesela** [comment]
Hey, I just want to add that although I am able to type diacritics with an international keyboard, it inserts the dead key that I press.
For example, pressing `~a` should type `à`, but it redundantly types `~ã`.
And for some info: I am on Windows with the latest version of Neovide from Chocolatey, and am using the windows US international keyboard layout.

**Comment by Kethku** [comment]
@lijaesela please try with current main. I think the chocolatey version may be old.

**Comment by Kethku** [comment]
Yeah I just confirmed that its built using the released version 0.7.0 which is definitely old.

**Comment by awarewen** [comment]
OS: arch linux 5.13.10-arch1-1 x86_64
WM: KDE plasma 5.22.4

I used fcitx5 can't input chinese.

**Comment by Kethku** [comment]
@awarewen that is blocked by better ime support upstream which is targeted to merge after the keyboard reworks

**Comment by kov** [comment]
I built master and have the same issue as @lijaesela but I'm on Linux, I assume that is still in the works for this platform? Maybe something missing in winit?

**Comment by Kethku** [comment]
Its possible something broke this recently :( Can you give some details about exactly what you input, what you expected, and what happened instead along with your keyboard layout?

**Comment by fredizzimo** [comment]
I recompiled with the latest to try to confirm what @lijaesela and @kov are reporting, but I don't see any problems with my Swedish keyboard layout on Windows, the dead keys works like they should. So it looks like at least the Windows version is still working as it should.

Except for a new thing that I just noticed. This is not a new regression, I just haven't noticed it before due to my mappings, and it might be the same issue as  https://github.com/neovide/neovide/issues/832. The problem is that I can't exit the terminal mode with `Ctrl-\ Ctrl-N` with my Swedish layout. On other layouts, like the US one it works fine.

The problem seems to be related to the control mappings, and they are very strange in the terminal and Vim see https://vimhelp.org/vim_faq.txt.html#faq-20.5 and https://vi.stackexchange.com/questions/5732/tag-navigation-using-ctrl-does-not-work-with-non-english-keyboard-on-windows. 

Basically in a terminal Vim, even if you have an international keyboard layout, you need to press the same physical key as the US layout for the corresponding key to generate the correct control code. In my case I have to press the `Ctrl-<` to generate `Ctrl-\`. 

But in Neovide this does not work. The combination `Ctrl-AltGr-+` does not work either, which would be the correct combination if the keymap is actually in use. 

I would personally prefer to not emulate the terminal in this case, so that the keys from the actual keymap generates the correct control codes. Then all ctrl combinations would probably be supported. But I'm sure many people would prefer the behaviour that the terminal Vim has, especially since those combinations are much easier to press than the `Ctrl-AltGr' combinations that would be needed otherwise.

**Comment by hemedani** [comment]
`super v` do not work in my system even when I bind it to the last register
```
vim.api.nvim_set_keymap("i", "<D-v>", "<C-r>+", { noremap = true, silent = true })
```

it's work in older version of nevoid

**Comment by hemedani** [comment]
also `alt` or `option` key do not work completely

**Comment by kov** [comment]
If I type ~<space> I expect a ~ to appear, but I get nothing. Also 'a should give me á, but nothing happens either, it looks like dead keys are not working at all under Linux.

**Comment by Kethku** [comment]
Linux: https://github.com/rust-windowing/winit/pull/1932
Mac: https://github.com/rust-windowing/winit/pull/1890

Although currently neovide is built upon https://github.com/Kethku/winit/tree/new-keyboard-all which is a branch which merges all three pull requests. I'm guessing @unclechu that your issue was fixed here: https://github.com/rust-windowing/winit/pull/1890#issuecomment-884811705. Are you running on a mac? If so, then I just need to cherry pick the upstream commit and we should be good to go.

**Comment by unclechu** [comment]
@Kethku nah, I’m on NixOS. I assume https://github.com/rust-windowing/winit/pull/1932 is not ready yet since it’s only a draft?

**Comment by 22h49** [comment]
Hey I've got a question, I use xkb-switch and I switch between my layouts, I've noticed that neovide doesn't pick this up.

How do I make it reload the current keyboard layout?

**Comment by unclechu** [comment]
@22h49 not possible at the moment. Read the whole thread. Neovide depends on **winit** which handles the keyboard input and it does not support any layouts in fact. It just interprets bare key codes as some keys from one its own qwerty-like layout. See one of the latest comments from this merge request: https://github.com/rust-windowing/winit/pull/1932#issuecomment-923221013.

**Comment by Kethku** [comment]
Thats not quite true. Neovide in main does pay attention to keyboard layouts. We use a branch which pulls those PRs in from mac windows and linux so that we can do better.

Its not surprising that there are still issues though.

**Comment by maroider** [comment]
The latest changes on the Linux PR should respect keyboard layout changes, but those changes don't seem like they've been added to https://github.com/Kethku/winit/tree/new-keyboard-all.

**Comment by Kethku** [comment]
OK thank you! I'll update the fork shortly

**Comment by imansour12** [comment]
I think the bug is still there, whenever i press AltGR and 3 to make a "#", it prints a four although the other text editors do it just fine.

**Comment by TobiasDev** [comment]
I have the same with {} on a swedish keyboard. They also don't work.

**Comment by fredizzimo** [comment]
@imansour12 and @TobiasDev, which version of Neovide are you using? I don’t think the fixes are in the 0.7 release, so you need to build it from the sources.

**Comment by LudoPinelli** [comment]
Just to mention that with the version built from source, everything seems to work just fine now using a french keyboard on Windows, but it still doesn't work if I launch neovide from wsl.

**Comment by imansour12** [comment]
@fredizzimo 
I realize, aplogies.
@LudoPinelli 
Hey mate, can you please send me the version you compiled? I would really appreciate that. email: iyedmansour0@gmail.com

**Comment by clason** [comment]
I strongly advise never to execute binaries you got via email from random people on the internet (nor, in fact, to send such).

Instead, if you're not able to compile yourself (which is not hard!), you can download binaries compiled by CI from the Actions page, e.g., https://github.com/neovide/neovide/actions/runs/1648397985

**Comment by imansour12** [comment]
@clason Cool cool cool, it works fine now! Thanks a lot mate!!

**Comment by HerCerM** [comment]
In macOS I use the **Spanish Latin American** keyboard layout, and I have found these issues for the dead keys `´` and `¨`:

Typing `´J` produces `´`, instead of the expected `´J`. Similarly, `¨J` produces `¨` instead of `¨J`.

> ℹ️ The issue for `¨` is specific when releasing Shift and pressing it again to input the uppercase letter. If Shift is held during the entire input, the issue does not happen.

I can reproduce this with any **uppercase letter**, for some reason with lowercase letters there is no issue:

Typing `´j` produces the expected `´j`. Similarly, `¨j` produces `¨j`.

This is a problem for my workflow since I use `´` and `¨` as a prefix for various mappings. In iTerm2 this works fine.

Would this inconsistency be solved by this issue? Thanks.

---

Additional info:

I installed Neovide using Homebrew, with `brew install --cask neovide`.

```text
$ neovide --version
Neovide 0.8.0
```

**Comment by Andy3153** [comment]
Still having problems. With both #492/#1073, using the Romanian (standard) layout inside KDE, but also with #503, where I have a keybind to `Ctrl + /` that calls commands from [numToStr/Comment.nvim](https://github.com/numToStr/Comment.nvim)

I am using the `neovide` AUR package.

``` sh
$ neovide --version

Neovide 0.8.0
```

**Comment by ShenMian** [comment]
Related issue: #1006.

**Comment by fredizzimo** [comment]
Hopefully the new Winit keyboard API has fixed all the remaining issues. It's now on Neovide master with https://github.com/neovide/neovide/pull/1789, so could you try with that?

**Comment by Andy3153** [comment]
@fredizzimo Just tried the master branch right now and it worked for me. I only tested for this specific problem, so I don't yet know if other problems arose though.

**Comment by fredizzimo** [comment]
I think if the issue @hemedani has is also fixed we should finally close this and open new issues for specific bugs if necessary. 

For IME support we have this https://github.com/neovide/neovide/issues/1127

**Comment by alerque** [comment]
I just compiled Git HEAD now that the winint stuff has been merged and was able to use my usual keyboard inputs (I use Programmer-Dvorak, Turkish-F, and a standard Russian layout). Switching layouts in Wayland while using Neovide works as expected.

I was *not* able to use the Compose key sequences I normally would in other apps.

**Comment by clason** [comment]
Yep, dead keys support (`´` + `e` -> `é`) stopped working (this is a recent regression).

**Comment by fredizzimo** [comment]
@clason, do you have more information about your system?

I use dead keys all the time with the Swedish keyboard layout, and for me it's working in Windows, Wayland and X11. Inculding `´` + `e`

**Comment by clason** [comment]
German keyboard layout, macOS. (Dead keys is _very_ platform dependent; we had to jump through a lot of hoops to get this working way back when.)

**Comment by fredizzimo** [comment]
We need to determine if it's as winit bug or Neovide bug then. So someone on MacOs has to debug it and figure it out.

Neovide in itself does not have any platform specific code as far as I can see https://github.com/neovide/neovide/blob/main/src/window/keyboard_manager.rs

**Comment by clason** [comment]
https://github.com/neovide/neovide/pull/767 added it.

If you give me pointers on what to check, I will do so. (I haven't followed recent development of Neovide, I have to admit...)

**Comment by fredizzimo** [comment]
Yes, but that platform specific code is not for the dead keys, just for the alt/meta key swapping.

The dead key handling was later refactored to the platform generic way it is now, and using the winit provided `key::Dead` instead of doing our own handling and that has been the case for a couple of years now.

**Comment by clason** [comment]
> Yes, but that platform specific code is not for the dead keys, just for the alt/meta key swapping.

No, the modifiers bit is relevant here as well (do look at the PR I linked).

> The dead key handling was later refactored to the platform generic way it is now, and using the winit provided key::Dead instead of doing our own handling and that has been the case for a couple of years now.

That is correct, but the last commit regressed on that. Looks like I will have to dive into the diff for that ...

**Comment by fredizzimo** [comment]
Ok, I looked a bit closer, on MacOS we call `text_with_all_modifiers` and on other platforms `key_without_modifiers`. My guess is that winit has fixed that platform difference, and now we should call `key_without_modifiers` on all platforms.

**Comment by clason** [comment]
No, that's not it (and it's also needed to handle alt-modified keys correctly).

I suspect the `map to_string` is stripping stuff out.

**Comment by fredizzimo** [comment]
> I suspect the map to_string is stripping stuff out.

That would be very weird, since we do that on all platforms, and it's not stripping anything out on those. And it just converts `Option<&str>` to `Option<String>`

The conversion to string is needed because `text_with_all_modifiers()` returns `Option<&str>`  while `KeyEvent.text` is `Option<SmolStr>`

**Comment by clason** [comment]
Let's move this to #1896

**Comment by clason** [comment]
I'm not fully convinced that these are completely different issues, but if you prefer we can do this.

**Comment by clason** [comment]
There's of course the very real (and even more likely) possibility that the issue is with upstream winit rather than any change here.

**Comment by fredizzimo** [comment]
That one was reverted here https://github.com/rust-windowing/winit/pull/2119

**Comment by clason** [comment]
Sorry, I missed that this got merged in the main repo. As I wrote, I have not been following Neovide and upstream lately, so I should also refrain from needless speculation.

TL;DR: I can only report that things are broken and am happy to test whatever you tell me to test, but will shut up otherwise.

**Comment by clason** [comment]
> You could try this commit https://github.com/neovide/neovide/commit/2a82eae6ac9171d783480d662037f18f8166660c

Same issue, unfortunately.

> You can also test alacritty with the winit master https://github.com/kchibisov/alacritty/tree/winit-update. And if that works, I guess we just have to compare our implementation to theirs.

That handles dead/composing keys correctly.

**Comment by fredizzimo** [comment]
> That handles dead/composing keys correctly.

Great. I prepare a PR that simplifies our input to something very similar. Not sure when I will have time do do it, but probably during the weekend. 

If someone else wants to do it, notify me before you start and I will do the same here. That way we don't duplicate the work.

**Comment by fredizzimo** [comment]
I started looking at using the API the same way as Alacritty now. It looks promising, the dead keys works at least on my system without any kind of special handling. 

The biggest thing I'm worried about though is the alt/meta swap on MacOS, since I don't have a Mac and the Alacritty branch has not implemented that yet, so I don't have any reference to look at. 

The Alacritty branch do have IME support, and I think I should be able to test that, so I should probably be able to get that working. 

I will also look at some of our mapping problems for example ctrl+number, if that's easily fixable now.

**Comment by clason** [comment]
The big thing (for me) related to that is third-layer characters like `\` -- which seemed to work on alacrity.

Happy to test things in any case.

**Comment by fredizzimo** [comment]
I created the very first draft version https://github.com/neovide/neovide/pull/1899.

I have personally only tested it on Wayland, but as far as I can see the dead keys and regular mappings work, even the ctrl-number ones that previously didn't.

I haven't attempted to implement `neovide_input_macos_alt_is_meta` and `neovide_input_use_logo` yet.

IME input is also missing.

Despite that I'm happy for all the testing I can get, especially on other platforms. I can only test Wayland, X11 and Windows myself, and working on Wayland currently so testing the other takes a bit of effort.

**Comment by unclechu** [comment]
My issue was in 2021:

> When I try to type for instance $ on a Finnish layout (which is [AltGr+4](https://upload.wikimedia.org/wikipedia/commons/3/35/KB_Finnish_Multilingual.svg)) it does something weird (goes to Normal mode).

I tested latest HEAD f412399676c83ea670cd22745b850dad9ee210a0 (see my Nix config for it https://github.com/unclechu/nixos-config/commit/a65b44528a24996bd6232b10bd610ada0b49c501 ) and it doesn’t seem to reproduce anymore.

**Comment by fredizzimo** [comment]
I marked #1899 as ready now. It should fix all the issues and also partially enable IME, which will properly be fixed in a later PR.

**Comment by unclechu** [comment]
I found out that `<A-...>` mappings are still prone to this kind of issue. Created a separate ticket for this: https://github.com/neovide/neovide/issues/1901
